### PR TITLE
Upgrade to SciMLOperators

### PR DESCRIPTION
**Pull request by vpuri3**
plan: https://github.com/SciML/SciMLOperators.jl/issues/142

To be clear, DiffEqOperators is not being deprecated in this PR. I am only adding support for `SciMLOperators.jl` by moving `AbstractSciMLOperator` and many of the traits to `SciMLOperators.jl` and making `DiffEqOperators` a subtype. That way, all the stuff in `SciMLOperators.jl` will be supported in downstream packages.

### CODE REVIEW

**Comment by vpuri3** [comment]
@ChrisRackauckas

**Comment by vpuri3** [comment]
oops looks like some tests failed

**Comment by vpuri3** [comment]
looks like none of the errors are related to the changes made here. @xtalax can you help?

**Comment by xtalax** [comment]
@ChrisRackauckas can confirm that the flux issue is present on master, excepting invalidations LGTM

**Comment by vpuri3** [comment]
Okay, I took a look at all the errors:

In `OrdinaryDiffEq.jl`, the test is failing because of low accuracy. However the offending `@test`, however, is comparing floats with `=` instead of `\approx`, and has no tolerance value.

https://github.com/SciML/OrdinaryDiffEq.jl/blob/2426acd6342a17add6e9fee979bf7f53d2d97270/test/regression/psos_and_energy_conservation.jl#L48

I reran that entire test group on my personal mac (with this branch of `SciMLBase.jl`), and the test passed. Then I did the same thing on a linux cluster and it got no errors. So this must be a numerical fluke.

Screenshot 2023-02-02 at 2 46 04 PM

For `DiffEqFlux`, the test is erroring on the `NeuralCDDE` function call. I am getting that error with when I test `DiffEqFlux` without this PR so this doesn't seem related to my work. 

https://github.com/SciML/DiffEqFlux.jl/blob/7e52962f011f1a9ebf39a05c0f48ffd3a9b659df/test/neural_de_lux.jl#L274

Screenshot 2023-02-02 at 2 45 55 PM

**Comment by vpuri3** [comment]
yay all tests have passed! @ChrisRackauckas this is good to go whenever you have time.

File: src/operators/operators.jl [hunk 1]
```diff
@@ -1,61 +1,39 @@
...
-
-Base.eltype(::Type{AbstractSciMLOperator{T}}) where {T} = T
-Base.eltype(::AbstractSciMLOperator{T}) where {T} = T
-update_coefficients!(L, u, p, t) = nothing
-update_coefficients(L, u, p, t) = L
+#
+#Base.eltype(::Type{AbstractSciMLOperator{T}}) where {T} = T
+#Base.eltype(::AbstractSciMLOperator{T}) where {T} = T
+#update_coefficients!(L, u, p, t) = nothing
+#update_coefficients(L, u, p, t) = L
```
Comment (ChrisRackauckas) [inline src/operators/operators.jl:5]: No commented stuff like this, please clean.

Comment (vpuri3) [reply src/operators/operators.jl]: done

**Comment by vpuri3** [comment]
anything else @ChrisRackauckas ?

File: src/deprecated.jl [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+const DiffEqIdentity = IdentityOperator
+function DiffEqIdentity(u)
+    @warn "SciMLBase.DiffEqIdentity is deprecated.
+    Use SciMLOperators.IdentityOperator instead"
+
+    IdentityOperator{size(u, 1)}()
+end
+
+const DiffEqScalar = SciMLOperators.ScalarOperator
+function DiffEqScalar(args...; kwargs...)
```
Comment (ChrisRackauckas) [inline src/deprecated.jl:48]: How does this work if this is happening? Add a downstream test onto LinearSolve.jl

Comment (vpuri3) [reply src/deprecated.jl:48]: DiffEqOperators is not being deprecated in this PR. These lines are all commented out. I am only adding support for `SciMLOperators.jl` by moving `AbstractSciMLOperator` and many of the traits to `SciMLOperators` and making `DiffEqOperators` a subtype. That way, all the stuff in `SciMLOperators` will be supported in downstream packages.
 Read more here: https://github.com/SciML/SciMLOperators.jl/issues/142

Comment (vpuri3) [reply src/deprecated.jl:48]: downstream test for what - i dont follow

Comment (ChrisRackauckas) [reply src/deprecated.jl:48]: > and making DiffEqOperators a subtype

I don't see that. I just see a bunch of comments.

File: src/deprecated.jl [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+# Deprecated operator interface
+#=
+const AbstractDiffEqOperator = AbstractSciMLOperator
+const AbstractDiffEqLinearOperator = AbstractSciMLOperator
+const AbstractDiffEqCompositeOperator = AbstractSciMLOperator
+
+const DiffEqScaledOperator = ScaledOperator
+function DiffEqScaledOperator(args...; kwargs...)
+    @warn "SciMLBase.DiffEqScaledOperator is deprecated.
```
Comment (ChrisRackauckas) [inline src/deprecated.jl:25]: Clean please

File: src/operators/common_defaults.jl [hunk 4]
```diff
@@ -1,6 +1,9 @@
 # The `update_coefficients!` interface
-DEFAULT_UPDATE_FUNC(A, u, p, t) = A # no-op used by the basic operators
+
+#DEFAULT_UPDATE_FUNC(A, u, p, t) = A # no-op used by the basic operators
```
Comment (ChrisRackauckas) [inline src/operators/common_defaults.jl:3]: clean.,..

Comment (vpuri3) [reply src/operators/common_defaults.jl]: done

File: src/deprecated.jl (hunk 3 above)
Comment (vpuri3) [reply src/deprecated.jl]: done

File: src/deprecated.jl (hunk 2 above)
Comment (vpuri3) [reply src/deprecated.jl]: fair. the comments have been removed. apologies for the mess and thanks for pointing to the issue. I've removed unnecessary comments for this PR.

**Comment by vpuri3** [comment]
@ChrisRackauckas anything else that needs to be done here, or are we just waiting on tests?

**Comment by ChrisRackauckas** [comment]
Add LinearSolve.jl to the downstream tests

**Comment by vpuri3** [comment]
hold on - just noticed `LinearSolve` was in downstream tests all along:

https://github.com/SciML/SciMLBase.jl/blob/f9e2da075eed9032fb886d847d3a1b6124a083bc/.github/workflows/Downstream.yml#L28-L37

**Comment by vpuri3** [comment]
And the test did pass. So I'll remove the line
line 28 - `- {user: SciML, repo: LinearSolve.jl, group: Core} `, because the test on line 37 ` - {user: SciML, repo: LinearSolve.jl, group: All} ` should be sufficient

**Comment by vpuri3** [comment]
So we should be good to go @ChrisRackauckas ?

**Comment by vpuri3** [comment]
ok LinearSolve test passed again:

```julia
      Testing LinearSolve tests passed 
 ```

https://github.com/SciML/SciMLBase.jl/actions/runs/4093861110/jobs/7059588503

**Comment by ChrisRackauckas** [comment]
> and making DiffEqOperators a subtype.

Where is this done?

**Pull request merged**